product_pack/models/product_template.py

```diff
@@ -0,0 +1,101 @@
+    )
+    pack = fields.Boolean(
+        'Pack?',
+        help='Is a Product Pack?',
+    )
+    pack_line_ids = fields.One2many(
+        related='product_variant_ids.pack_line_ids',
+        readonly=True,
+    )
+    used_pack_line_ids = fields.One2many(
+        related='product_variant_ids.used_pack_line_ids',
+        readonly=True,
+    )
+    allow_modify_pack = fields.Boolean()
+
+    @api.onchange('pack_type')
+    def onchange_allow_modify_pack(self):
+        products = self.filtered(
+            lambda x: x.allow_modify_pack and
+            x.pack_type and x.pack_type != 'components_price')
+        for rec in products:
+            rec.allow_modify_pack = False
+
+    @api.constrains('pack_type')
+    def check_relations(self):
+        """
+        Check assisted packs don't have packs a child
+        """
+        for rec in self:
+            if rec.pack_type == 'none_detailed_assisted_price':
+                child_packs = rec.mapped(
+                    'pack_line_ids.product_id').filtered('pack')
+                if child_packs:
+                    raise ValidationError(_(
+                        'A "None Detailed - Assisted Price Pack" can not have '
+                        'a pack as a child!'))
+            if rec.pack:
+                used_lines = rec.product_variant_ids.used_pack_line_ids
+                for product in used_lines.mapped('parent_product_id'):
+                    if product.pack_type == 'none_detailed_assisted_price':
+                        raise ValidationError(_(
+                            'You can not set this product as pack because it '
+                            'is part of a "None Detailed - Assisted Price '
+                            'Pack"'))
+
+    @api.constrains('company_id', 'product_variant_ids')
+    def check_pack_line_company(self):
+        """
+        Check packs are related to packs of same company
+        """
+        for rec in self:
+            for line in rec.pack_line_ids:
+                if (line.product_id.company_id and rec.company_id) and \
+                        line.product_id.company_id != rec.company_id:
+                    raise ValidationError(_(
+                        'Pack lines products company must be the same as the '
+                        'parent product company'))
+            for line in rec.used_pack_line_ids:
+                if (line.product_id.company_id and rec.company_id) and \
+                        line.parent_product_id.company_id != rec.company_id:
+                    raise ValidationError(_(
+                        'Pack lines products company must be the same as the '
+                        'parent product company'))
+
+    @api.multi
+    def write(self, vals):
+        """
+        We remove from prod.prod to avoid error
+        """
+        if vals.get('pack_line_ids', False):
+            self.product_variant_ids.write(
+                {'pack_line_ids': vals.pop('pack_line_ids')})
```

Comment:
:warning: :fearful: This code triggers some internal alerts. What is this supposed to be doing? Are you conscientious it's altering the dict? Is it on purpose?

Reply:
Yes, I'm conscientious it's altering the dict, the purpose of who did this in previous versions was to be able to manage pack lines of product.product from product.template (when there are no variants). I kept it in that way because I didn't see that it could cause any problem. Do you think that would affect something else?

Reply:
@Yajo , I don't understand step 1. Anyway, what I did was remove the `readonly = true` that has no effect because in the product.template view is set readonly again with a condition. I also used a _vals variable as you state in step 3 to avoid unexpected alterations.

Reply:
He is asking for a video showcasting what this piece of code is doing.